## **Data Quality Assessment and Data Cleaning: Customer Address:**

## **1. Importing Libraries:**

In [1]:
import pandas as pd 
import numpy as np 
import matplotlib.pyplot as plt 
%matplotlib inline 
import seaborn as sns 

from datetime import datetime, date 

## **2. Data Ingestion:**

In [3]:
# Load the customer address data from the Excel file 
cust_address = pd.read_excel("Raw_data.xlsx", sheet_name = "CustomerAddress") 
cust_address.head(3) 

customer_id              address  postcode            state    country  \
0            1   060 Morning Avenue      2016  New South Wales  Australia   
1            2  6 Meadow Vale Court      2153  New South Wales  Australia   
2            4   0 Holy Cross Court      4211              QLD  Australia   

   property_valuation  
0                  10  
1                  10  
2                   9

In [5]:
# Let's check information of columns and data-types of Customer Address data 
cust_address.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3999 entries, 0 to 3998
Data columns (total 6 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   customer_id         3999 non-null   int64 
 1   address             3999 non-null   object
 2   postcode            3999 non-null   int64 
 3   state               3999 non-null   object
 4   country             3999 non-null   object
 5   property_valuation  3999 non-null   int64 
dtypes: int64(3), object(3)
memory usage: 187.6+ KB


### Rows in the DataFrame:

The DataFrame consists of 3999 rows, each representing a unique customer with specific details such as customer ID, address, postal code, state, country, and property valuatio



## **3. Data Exploration:**

In [8]:
# Let's check total records and features in the dataset
cust_address.shape

(3999, 6)

In [9]:
# Numeric columns 
df_numeric = cust_address.select_dtypes(include = [np.number]) 
df_numeric.head(3) 

customer_id  postcode  property_valuation
0            1      2016                  10
1            2      2153                  10
2            4      4211                   9

In [12]:
# Check numeric columns 
print(df_numeric.columns)

Index(['customer_id', 'postcode', 'property_valuation'], dtype='object')


In [14]:
# Select non-numeric columns 
df_non_numeric = cust_address.select_dtypes(exclude = [np.number]) 
print(df_non_numeric.columns) 

Index(['address', 'state', 'country'], dtype='object')


In [15]:
df_non_numeric.head(3) 

address            state    country
0   060 Morning Avenue  New South Wales  Australia
1  6 Meadow Vale Court  New South Wales  Australia
2   0 Holy Cross Court              QLD  Australia

In [16]:
# Missing values check 
cust_address.isnull().sum()

customer_id           0
address               0
postcode              0
state                 0
country               0
property_valuation    0
dtype: int64

In [17]:
# Inconsistency check in data 
cust_address["state"].value_counts()

NSW                2054
VIC                 939
QLD                 838
New South Wales      86
Victoria             82
Name: state, dtype: int64

#### **Observation:**

**To address inconsistent data in the `'State'` column:**

- Issue: Variations in 'State' column entries for 'New South Wales' and 'Victoria' (full names and short names).
- Standardization: Replace 'New South Wales' with 'NSW.', Replace 'Victoria' with 'VIC.'
- Implementation: Utilize pandas to modify the 'State' column. Replace occurrences as specified.
- Result: Achieve consistency by using 'NSW' for 'New South Wales' and 'VIC' for 'Victoria' in the 'State' column.
- This ensures standardized state representation.To address inconsistent data in the 'State' column:
- **Implementation:** Utilize pandas to modify the 'State' column. Replace occurrences as specified.
- **Result:** Achieve consistency by using 'NSW' for 'New South Wales' and 'VIC' for 'Victoria' in the 'State' column. This ensures standardized state representation.

In [18]:
# Function to replace full state names with their short forms 
def replace_state_names(state_name):
    """
    Making short form of state names as standards
    """
    if state_name == "New South Wales":
        return "NSW"
    elif state_name == "Victoria":
        return "VIC"
    else:
        return state_name 

# Apply the above function to state column 
cust_address["state"] = cust_address["state"].apply(replace_state_names)

In [19]:
# Let's check the result now 
cust_address["state"].value_counts()

NSW    2140
VIC    1021
QLD     838
Name: state, dtype: int64

### **Note:**

- After applying the above function, the `state` name is **`standardized`**.
- Now, there is no inconsistency in the state column.

In [20]:
# Let's check country column
cust_address["country"].value_counts()

Australia    3999
Name: country, dtype: int64

In [22]:
# Let's check postcode column 
cust_address["postcode"].head()

0    2016
1    2153
2    4211
3    2448
4    3216
Name: postcode, dtype: int64

In [24]:
cust_address[["address", "postcode", "state", "country"]]

address  postcode state    country
0            060 Morning Avenue      2016   NSW  Australia
1           6 Meadow Vale Court      2153   NSW  Australia
2            0 Holy Cross Court      4211   QLD  Australia
3           17979 Del Mar Point      2448   NSW  Australia
4              9 Oakridge Court      3216   VIC  Australia
...                         ...       ...   ...        ...
3994            1482 Hauk Trail      3064   VIC  Australia
3995  57042 Village Green Point      4511   QLD  Australia
3996     87 Crescent Oaks Alley      2756   NSW  Australia
3997           8194 Lien Street      4032   QLD  Australia
3998            320 Acker Drive      2251   NSW  Australia

[3999 rows x 4 columns]

In [25]:
cust_address.head(5)

customer_id              address  postcode state    country  \
0            1   060 Morning Avenue      2016   NSW  Australia   
1            2  6 Meadow Vale Court      2153   NSW  Australia   
2            4   0 Holy Cross Court      4211   QLD  Australia   
3            5  17979 Del Mar Point      2448   NSW  Australia   
4            6     9 Oakridge Court      3216   VIC  Australia   

   property_valuation  
0                  10  
1                  10  
2                   9  
3                   4  
4                   9

In [27]:
cust_address[["address", "postcode", "state", "country"]].drop_duplicates()

address  postcode state    country
0            060 Morning Avenue      2016   NSW  Australia
1           6 Meadow Vale Court      2153   NSW  Australia
2            0 Holy Cross Court      4211   QLD  Australia
3           17979 Del Mar Point      2448   NSW  Australia
4              9 Oakridge Court      3216   VIC  Australia
...                         ...       ...   ...        ...
3994            1482 Hauk Trail      3064   VIC  Australia
3995  57042 Village Green Point      4511   QLD  Australia
3996     87 Crescent Oaks Alley      2756   NSW  Australia
3997           8194 Lien Street      4032   QLD  Australia
3998            320 Acker Drive      2251   NSW  Australia

[3999 rows x 4 columns]

#### **Duplication Checks:**

- We must ensure that there is no duplication of records in the dataset.
- Duplication may lead to error in data analysis due to poor data quality.
- If there are duplicate rows of data, then we need to drop such records.
- For checking duplicate records we need to firstly remove the primary key column of the dataset.
- Then apply `drop_duplicate()` method provided by Pandas DataFrame. 

In [28]:
# Drop the customer_id column and store it in a temporary dataframe 
cust_address_dedupped = cust_address.drop("customer_id", axis = 1).drop_duplicates()
cust_address_dedupped.shape

(3999, 5)

#### **Observation:**

- There are no duplicate records since both the numbers are same.

## **4. Export the cleaned customer demographic data set to csv file:**

In [29]:
cust_address.to_csv("CustomerAddress_Cleaned.csv", index = False) 

In [33]:
# Let's check cleaned customer address: 
df = pd.read_csv("CustomerAddress_Cleaned.csv")
df.head(2)

customer_id              address  postcode state    country  \
0            1   060 Morning Avenue      2016   NSW  Australia   
1            2  6 Meadow Vale Court      2153   NSW  Australia   

   property_valuation  
0                  10  
1                  10

## **5. Check of master-detail record counts:**

In [31]:
cust_demo_detail = pd.read_csv("CustomerDemographic_Cleaned.csv") 
cust_demo_detail.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3912 entries, 0 to 3911
Data columns (total 13 columns):
 #   Column                               Non-Null Count  Dtype  
---  ------                               --------------  -----  
 0   customer_id                          3912 non-null   int64  
 1   first_name                           3912 non-null   object 
 2   last_name                            3912 non-null   object 
 3   gender                               3912 non-null   object 
 4   past_3_years_bike_related_purchases  3912 non-null   int64  
 5   DOB                                  3912 non-null   object 
 6   job_title                            3912 non-null   object 
 7   job_industry_category                3912 non-null   object 
 8   wealth_segment                       3912 non-null   object 
 9   deceased_indicator                   3912 non-null   object 
 10  owns_car                             3912 non-null   object 
 11  tenure                        

In [35]:
cust_demo_detail.shape, cust_address.shape

((3912, 13), (3999, 6))

#### **Observation:**

- Total records in Customer Demographic Dataset is 3912.
- Total records in Customer Address Table is 3999.
- In demographic table, 87 records are getting dropped due to the data cleaning process. 

In [42]:
list(cust_address.columns)

['customer_id',
 'address',
 'postcode',
 'state',
 'country',
 'property_valuation']

In [41]:
list(cust_demo_detail.columns)

['customer_id',
 'first_name',
 'last_name',
 'gender',
 'past_3_years_bike_related_purchases',
 'DOB',
 'job_title',
 'job_industry_category',
 'wealth_segment',
 'deceased_indicator',
 'owns_car',
 'tenure',
 'Age']

In [45]:
# Merging cust_address and cust_demo_detail 
cust_drop = cust_address.merge(
    cust_demo_detail, 
    left_on = 'customer_id', 
    right_on='customer_id',
    how='outer')
cust_drop.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4003 entries, 0 to 4002
Data columns (total 18 columns):
 #   Column                               Non-Null Count  Dtype  
---  ------                               --------------  -----  
 0   customer_id                          4003 non-null   int64  
 1   address                              3999 non-null   object 
 2   postcode                             3999 non-null   float64
 3   state                                3999 non-null   object 
 4   country                              3999 non-null   object 
 5   property_valuation                   3999 non-null   float64
 6   first_name                           3912 non-null   object 
 7   last_name                            3912 non-null   object 
 8   gender                               3912 non-null   object 
 9   past_3_years_bike_related_purchases  3912 non-null   float64
 10  DOB                                  3912 non-null   object 
 11  job_title                     